### PR TITLE
🎉 Source Facebook Marketing: Added support for multiple Facebook accounts

### LINKED ISSUE #7740: Source Facebook Marketing: Support multiple Facebook accounts
## Tell us about the problem you're trying to solve
The Facebook Marketing source currently requires for one and only one account to be provided during the initial configuration. If you manage multiple Facebook Marketing accounts, it would be extremely handy to be able to sync data from many or even from all available Facebook accounts, not just one.

## Describe the solution you’d like
In the configuration, `account_id` should support multiple values and/or no values at all (to trigger a sync of all available accounts).

## Describe the alternative you’ve considered or used
An alternative would be to configure a source for each available account. Agencies can manage dozens (or even hundreds) of different accounts. This process would be impractical the more accounts you manage. If new accounts are added and removed on a regular basis, maintaining such a list becomes chaotic.

## Are you willing to submit a PR?
I will have a PR to share shortly and would greatly appreciate help and feedback making sure the changes are solid.

### PR DESCRIPTION
**Pull request by mlavoie-sm360**
## What
*Closes airbytehq/airbyte#7740

## How
- Updated the spec to allow one of two strategies regarding Facebook accounts:
  - All accounts linked to access token
  - One or many accounts as specified

## Recommended reading order
1. `spec.json`
2. `common.py`
3. `api.py`
4. `source.py`
5. `streams.py`
6. `async_job.py`

## 🚨 User Impact 🚨
Are there any breaking changes? If yes, please make sure to include it here and in any changelogs with the 🚨🚨 emoji
What is the end result perceived by the user?

- 🚨 The spec definition has been modified
- The end user will gain support for syncing all Facebook accounts associated to token

## Pre-merge Checklist

<details><summary> <strong> Updating a connector </strong></summary>
<p>
   
#### Community member or Airbyter
   
- [x] Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [x] Secrets in the connector's spec are annotated with `airbyte_secret` 
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated 
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] Changelog updated in `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
- [x] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
   
#### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items. 
   
- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] Credentials added to Github CI. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci). 
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing. 
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)
- [ ] After the new connector version is published, connector version bumped in the seed directory as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
- [ ] Seed specs have been re-generated by building the platform and committing the changes to the seed spec files, as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)

</p>
</details>

---
This change is [Reviewable](https://reviewable.io/reviews/airbytehq/airbyte/8237)

### CODE REVIEW

**Comment by marcosmarxm** [comment]
@mlavoie-sm360 please run `./gradlew format` to format all files?

File: airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/api.py
Comment (keu) [inline airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/api.py:109]: why do we need TTL here?

File: airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/api.py
Comment (keu) [inline airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/api.py:121]: what will happen if we can't find some accounts?

File: airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/spec.json
Comment (keu) [inline airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/spec.json:1]: what is the reason to roll back to spec.json ?

File: airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/common.py
Comment (keu) [inline airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/common.py:22]: this feels like re-inventing the wheel, is there any particular reason to replace pydantic with this?

File: airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/streams.py
Comment (keu) [inline airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/streams.py:230]: what is the motivation behind these changes?

File: airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/streams.py
Comment (keu) [inline airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/streams.py:203]: please don't use nested functions, it is a very bad practice, this affects readability and testing

Comment (keu) [reply airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/streams.py:203]: also could please ensure that any function has a docstring explaining why do we need it

File: airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/streams.py
Comment (keu) [inline airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/streams.py:202]:
```suggestion
        for account_id in stream_state.keys():
```

File: airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/common.py
Comment (mlavoie-sm360) [reply airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/common.py:22]: See thread in slack: https://airbytehq.slack.com/archives/C019WEENQRM/p1636488040184300

Summary: I am unable to create a "oneOf" type structure using pydantic. Any help welcome.

File: airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/spec.json
Comment (mlavoie-sm360) [reply airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/spec.json:1]: See previous comment on "oneOf" type structure

File: airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/streams.py
Comment (mlavoie-sm360) [reply airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/streams.py:230]: When syncing using the "all accounts" strategy, we could have hundreds of accounts to sync (it is our case, 500+), using this approach, we fetch and treat records one account at a time and leverage the `@backoff_policy` on `_read_records()`.

File: airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/api.py
Comment (mlavoie-sm360) [reply airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/api.py:121]: - If the api raises an exception, that exception is forwarded, same as before. 
- If you are using the subset strategy and some of the specified accounts are not found, those accounts are not included in list. 
- If you are using the subset strategy and all of the accounts are not found, an empty list of accounts is returned. 
- If you are using the all strategy, all found accounts are returned.

I'm not sure I'm understanding the question, am I missing something?

Comment (mlavoie-sm360) [reply airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/api.py:121]: Not finding some accounts raised an Exception. Now that we handle one or many accounts, I'm not sure what the behaviour should be. If I provide 5 accounts and I don't have access to one of them, do I want to raise an exception and break the flow, log a warning one of the accounts is missing or simply let the list have 4 accounts? I'm not sure what is the cleanest behaviour.

File: airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/spec.json
Comment (mlavoie-sm360) [reply airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/spec.json:1]: Waiting on the merge of PR [8485](https://github.com/airbytehq/airbyte/pull/8485) to implement a pydantic `oneOf` type structure.

File: airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/api.py
Comment (keu) [reply airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/api.py:121]: @mlavoie-sm360 
I think the check should succeed only if all specified accounts are accessible.

File: airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/common.py
Comment (keu) [reply airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/common.py:22]: I see, this PR should definitely help https://github.com/airbytehq/airbyte/pull/8485

File: airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/streams.py
Comment (keu) [reply airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/streams.py:231]: I mean the only change I see here is an indentation, so I was curious

Comment (mlavoie-sm360) [reply airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/streams.py:231]: Makes sense, to get a better picture, also check what changed in the different `def _read_records()` methods. If you configure a single account, the outcome is the same. If you configure multiple accounts, this change batch processes one account at a time.

File: airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/api.py
Comment (mlavoie-sm360) [reply airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/api.py:121]: @keu, thanks for your input. Is it all the same to you if that check is done in `source.py` under `def check_connection()` rather than here? At the moment, `check_connection` fails if:
- strategy is `subset` and at least one account is missing from the list
- strategy is `all` and no accounts are available
- strategy selected is unsupported

Comment (keu) [reply airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/api.py:121]: @mlavoie-sm360 I agree with the above if the check is done in the `source.py`, my only concern is that how the user will know that some accounts are not accessible anymore? How about we will have a warning for each missing account in `_find_account` method?

Comment (mlavoie-sm360) [reply airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/api.py:121]: @keu, how about this?

File: airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/source.py
Comment (keu) [inline airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/source.py:166]: @mlavoie-sm360 I think you forgot to revert this as well

File: airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/streams.py
Comment (keu) [inline airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/streams.py:300]: @mlavoie-sm360 I think this should be also updated to support multiple accounts, right?

File: airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/streams.py
Comment (keu) [inline airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/streams.py:295]: I see many repetitions of this, naturally CDK use the concept of `stream slices` to express loops in read_records. We can have default stream_slices methods looks like this:
```python
def stream_slices(....):
    for account in self._api.accounts:
        yield {'account': account}
```
then 
`request_params` method will take into account slice, no need to have `partition_field`.

**Comment by marcosmarxm** [comment]
@mlavoie-sm360 are you still working on this?

**Comment by mlavoie-sm360** [comment]
@marcosmarxm, I am, away from work at the moment, but I'll be picking up again. Feel free to pitch in if you want.

**Comment by marcosmarxm** [comment]
@mlavoie-sm360 do you think this is ready to review again?

**Comment by mlavoie-sm360** [comment]
@marcosmarxm, yes, ready for review

**Comment by marcosmarxm** [comment]
@keu can you review it again?

**Comment by keu** [comment]
> @keu can you review it again?

LGTM, but this one is blocked by the release of #9805 
after release I will update this PR with changes from master

**Review by marcosmarxm — APPROVED** [review]

**Comment by marcosmarxm** [comment]
@keu now it's possible to merge this?

**Comment by keu** [comment]
> @keu now it's possible to merge this?

@marcosmarxm this PR will need to merge with the latest master (will do it myself), currently working on #6225.

**Comment by marcosmarxm** [comment]
I'm closing this due inactivity.

**Comment by misteryeo** [comment]
**Update:** First off, thank you for your contribution @mlavoie-sm360, we really appreciate it! We're going to hold off reviewing this as it looks like this would require a non-trivial migration of configuration. This is something we're scheduled to implement this coming quarter which you can follow along here https://github.com/airbytehq/airbyte/issues/5396. Once complete, we'll come back and make sure we can get this merged.

Thanks for your patience!

**Comment by mlavoie-sm360** [comment]
Alright, thanks for the follow-up @misteryeo

**Comment by martiningolingo** [comment]
Hey, everyone! This would be a great improvement in the connection. 
I'm wondering when this can be available in the official release. 
Thank you for your work!

**Comment by jagannathsrs** [comment]
Is it `non-trivial migration` in the previous comment referring to existing pipelines? Will there be any issues if they create a new pipeline on this version?

**Comment by mlavoie-sm360** [comment]
@jagannathsrs, if you are creating a new pipeline on this version, there are no issues. The configuration is straight forward, the major difference being that you are not restricted to a single Facebook account.

**Comment by jagannathsrs** [comment]
> o

Awesome, I've cloned your repo and ran into this error:

```
The connection tests failed.
"FacebookAPIException('Error: 2635, (#2635) You are calling a deprecated version of the Ads API. Please update to the latest version: v13.0.')"
```

I don't see the API_VERSION configured anywhere in the code, is this set on the client end when generating the token?

File: airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/api.py
Comment (n0rritt) [inline airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/api.py:16]: must be `ConnectorConfig`  instead of `SourceFacebookMarketingConfig`

File: airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/api.py
Comment (n0rritt) [inline airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/api.py:127]: must be `self.__config.accounts.ids`

File: airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/api.py
Comment (n0rritt) [inline airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/api.py:145]: must be `self.__config.accounts.ids`

**Review by t0hai — CHANGES_REQUESTED** [review]
@mlavoie-sm360 could you fix the mentioned points? then your branch works for me in production and others might want to use your solution since FB just deprecated API v.12

**Comment by marcosmarxm** [comment]
Hello :wave:, first thank you for this amazing contribution. 

We really appreciate the effort you've made to improve the project.
We ask you patience for the code review. Last month our team was focused on Hacktoberfest event and that probably left some PR without the proper feedback. And this week, due to the Thanksgiving US Holiday, most our team is out of office with their families. Another important piece of information why code won't be merge this week is: as a safety measure the core team has decided to freeze merging code to main branch to keep the release stable. Next week we'll return to you with the proper code review and update the status of your contribution.

If you have any questions feel free to send me a message in Slack!
Thanks!

**Comment by marcosmarxm** [comment]
Hello :wave::skin-tone-2: and thank you for your contribution!

Airbyte has instituted a code freeze between 19 and 30 December, to make sure there are no disruptions during the holidays. 
Because of this, reviewing and merging your contribution may take longer than usual. 
We apologize for the delay, but we want everyone to have a quiet and happy holiday.

If you have any questions or need further clarification, please don't hesitate to ping via Slack.

**Comment by sherifnada** [comment]
Closing for inactivity -- will reopen or reimplement once the blocking config migration issue #5396 is resolved

**Pull request closed without merging**
